### PR TITLE
Typescript

### LINKED ISSUE #142: Typescript definitions
I'd love to see official typescript definitions, preferably shipped with the package. 

Fully typing `as` will be tricky - I tried my hand at it with little success, but it felt like it should be possible.

### PR DESCRIPTION
**Pull request by wordofchristian**
closes #142 

## Background

The goal of this branch is to add support for typescript so that it can be incrementally added to the library, eventually the library will be rewritten entirely in typescript but we don't want to do it all in one go to avoid conflicts. It is currently configured to type check javascript files in addition to typescript files. It should be possible to [fix type errors in the javascript files either by using jsdoc comments ](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html) or `@ts-ignore` statements.

I'll continue working on this  but the initial impressions look promising, the tests are currently passing and there are only 15 type errors left.

## Removed support for `babel-plugin-transform-export-extensions`

The syntax for `export default from "./Base";` is not yet supported by typescript and [won't be until it reaches stage 3 status](https://github.com/Microsoft/TypeScript/issues/4813#issuecomment-369742110)

Without this the `type-check` command does not understand that syntax and I couldn't find a good way to add babel plugins before the type-check command. It seemed like the lowest cost solution was to just stop using this syntax.

## TODO
- [x] Fix remaining type errors
- [x] Audit tsconfig and make sure it's correct and nothing essential is missing
- [x] [Add module declaration file generation to build/publish process](https://github.com/Microsoft/TypeScript-Babel-Starter#set-up-your-build-tasks)
- [x] Add `yarn run type-check` to CI
- [x] fix deploy previews
- [x] fix CI
- [x] move custom-typings.d.ts to `typings/npm/styled-components/`
- [x] convert a public module to typescript, probably a component, or better yet the `as` enhancer.
- [x] Fix the issue with "flat" vs "flatten"
- [x] Make it work with styled-components 3.4

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    "jsx": "react",
+    "declaration": true,
+    "outDir": "lib",
+    "strict": true,
+    "noImplicitAny": false,
+    "allowSyntheticDefaultImports": true,
+    "esModuleInterop": true,
+  },
+  "exclude": [
+    "website/**/*"
```
Comment (foxyblocks) [inline tsconfig.json]: TODO look at removing this

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    "baseUrl": ".",
+    "allowJs": true,
+    "checkJs": true,
+    "target": "esnext",
+    "module": "commonjs",
+    "jsx": "react",
+    "declaration": true,
+    "outDir": "lib",
+    "strict": true,
+    "noImplicitAny": false,
```
Comment (foxyblocks) [inline tsconfig.json]: TODO look at removing this, it adds a lot of type errors but ideally we could make new files in strict mode and manually disable errors in existing files.

File: website/src/utils/getSections.js [hunk 3]
```diff
@@ -1,3 +1,3 @@
-const getSections = section => [...section.sections, ...section.components];
+const getSections = section => [].concat(section.sections, section.components);
```
Comment (foxyblocks) [inline website/src/utils/getSections.js]: The previous syntax here was causing test failures when `section.sections` was undefined. I think the spread syntax may have changed subtly to disallow this `[...undefined]`.

File: src/custom-typings.d.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+/*
+  This file is to fix a missing type from styled-components.
+  It can be deleted when https://github.com/styled-components/styled-components/pull/1827 gets released
+*/
+import * as React from "react";
+import { Component } from "react";
+
+declare module "styled-components" {
+   export function isStyledComponent(target: string | Component<object>): boolean;
+}
```
Comment (foxyblocks) [inline src/custom-typings.d.ts]: Not crazy about this file name, it should probably be called `styled-components.d.ts`. I'm also not crazy about putting in directly in the `src` folder. 

Should look at moving it to a [`typings` folder](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html#types-typeroots-and-types). 

Maybe something like `typings/npm/styled-components/index.d.ts`

**Comment by diegohaz** [comment]
We probably have to upgrade (or resolve) styleguidist babel-loader to "babel-loader@^8.0.0-beta" in order to fix the website build.

**Comment by wordofchristian** [comment]
There seems to be an issue with singel in the cli when babel-register is loaded. I think it is trying to load the babel version for the project root rather than the one that singel needs.

**Comment by dicarlo2** [comment]
Adding `"babel-core": "7.0.0-bridge.0",` to the resolutions field in package.json might help, assuming the error has something to do with using `babel-core` vs `@babel/core`

**Comment by wordofchristian** [comment]
> Adding "babel-core": "7.0.0-bridge.0", to the resolutions field in package.json might help, assuming the error has something to do with using babel-core vs @babel/core

@dicarlo2 

I tried it with this:
```
  "devDependencies": {
    "@babel/cli": "7.0.0-beta.53",
    "@babel/core": "^7.0.0-beta.53",
// ...
    "babel-core": "^7.0.0-bridge.0",
    "babel-register": "^7.0.0-beta.3",
  },
  "resolutions": {
    "singel/babel-core": "7.0.0-bridge.0",
    "singel/babel-register": "^7.0.0-beta.3"
  }
```

and I still get this error:

```
$ singel src/components/**/[A-Z]*.js -i 'src/**/*{test,Container}.js'

/Users/crisco/code/reakit/node_modules/babel-register/node_modules/babel-core/lib/config/option-manager.js:122
      throw e;
      ^

Error: [BABEL] /Users/crisco/code/reakit/src/components/Arrow/Arrow.js: Requires Babel "^7.0.0-beta.41", but was loaded with "7.0.0-beta.3". You'll need to update your @babel/core version.
```

That's as close to the version as I can get. `babel-register` only goes up to version `7.0.0-beta.3`

**Comment by diegohaz** [comment]
I'm gonna checkout to this branch and test it later today. Thank you for your amazing work, @wordofchristian.

**Comment by wordofchristian** [comment]
I haven't gotten to the step of linking it and trying it out in a new project, let me know if it works for you.

**Comment by wordofchristian** [comment]
> .ts files weren't being linted, so I tried to integrate typescript with eslint using things like eslint-plugin-typescript, but that didn't work. We'll probably have better luck when babel/babel-eslint#505 lands.
> 
> I've added tslint then and played a lot with the code. Tried to add types to src/enhancers/as.js without success so far, probably because I'm still not very familiar with typescript. TS errors aren't clear enough to me (feels like Flow 😅).

I also took a stab at adding some types to `as`. It's going to require a bunch of generics I'm sure to get it fully finished. I've been looking at the type definitions for styled components to get some ideas, but their declaration file is pretty dense and unclear what everything is supposed to be doing. 

tslint seems like it might be the way to go, but it would be unfortunate to have to maintain two linter configs.

**Comment by diegohaz** [comment]
> but it would be unfortunate to have to maintain two linter configs.

Agreed. Maybe I did it wrong. This repository seems to work:
https://github.com/azdanov/tslint-eslint-crats#eslint--tslint

**Comment by diegohaz** [comment]
Maybe this can help us: https://github.com/neotracker/neotracker/blob/d4c9acc0538f16c5b07f57226965ba02a2645baf/%40types/reakit/index.d.ts

**Comment by wordofchristian** [comment]
> Maybe this can help us: https://github.com/neotracker/neotracker/blob/d4c9acc0538f16c5b07f57226965ba02a2645baf/%40types/reakit/index.d.ts

Lol, yeah this will help. I wonder why they didn't submit it as PR?

**Comment by Thomazella** [comment]
> Maybe this can help us: https://github.com/neotracker/neotracker/blob/d4c9acc0538f16c5b07f57226965ba02a2645baf/%40types/reakit/index.d.ts> 

😮

**Comment by dicarlo2** [comment]
Sorry, it's been a busy couple of weeks - I meant to point you guys to those typings in case you wanted to use them. However, they currently do not work with the latest release of typescript 3.0 and I haven't had a chance to look into why (at least in our repo they don't work). They do work on `3.1.0-dev.20180717` in case you want to play around with them.

One thing to note is that the typings are extremely abusive of typescript's type system, using hacky workarounds for recursive types in order to type `as` correctly (both the functional and prop form). They're likely to break for various reasons between typescript releases as a result. With that said, they're very specific, so, for example:
```typescript
<Base href="http://reakit.io" /> // Type error, href is not a valid property for Base or a div
<Base<'a'> as="a" href="http://reakit.io" /> // Works
```
Due to limitations in typescript's inference of literal types, the above typings sometimes require explicit hints (e.g. `Base<'a'>` above). I haven't gotten around to modeling `as` with multiple intrinsic overrides, but it should be possible as well (e.g. `Base<'a'>` is equivalent to the props of `Base`, `a` and `div` when it should only be `Base` and `a`).

**Comment by diegohaz** [comment]
- I've managed to configure eslint to work with typescript and removed tslint altogether.
- Put `styled.ts` in enhancers.
- Created `.vscode/settings.json` so eslint extension will work with typescript files too;
- Made some adjustments on `as.js`, but I'm not 100% sure about what I did there haha

**Comment by wordofchristian** [comment]
> I've managed to configure eslint to work with typescript and removed tslint altogether.
sweet!

> Made some adjustments on as.ts, but I'm not 100% sure about what I did there haha

Mostly looks like whitespace changes.


I won't have to work on this more today, but tomorrow for sure.

**Comment by wordofchristian** [comment]
Does this look interesting https://github.com/gcanti/prop-types-ts, in order to combine runtime and static type checking?

**Comment by codyaverett** [comment]
@wordofchristian @diegohaz how is this going?  Do you need help with anything?

Perhaps Typewiz could help this effort? `Automatically discover and add missing types in your TypeScript code`
https://github.com/urish/typewiz

**Comment by wordofchristian** [comment]
Hello, yes, sorry about the slow progress. We are close to getting the core pieces converted to typescript, those are the `as` helper and the `Base` component. Almost everything else is built on top of these so once they are in place the rest of the type conversion should go pretty smoothly.

I did successfully get it working in the latest commit, but it was only working with styled-components type definitions from `3.3.3` and broke again with `3.4`, probably not anything too complicated to figure out.


> Typing As looks both complicated and interesting.
> When I try to run the dev server on this branch I am getting [ts] Property 'flat' does not exist on type 'AsComponents[]'. This is a little over my head right now.

yes, there is currently a discrepancy between the typescript version and the babel polyfill. I think babel is trying to use the old name for this method `flatten()` where typescript is checking for the new name `flat()`, so that's one another item that needs to be address, probably just requires a babel tweak. 


Another thing I want to figure out but is not a blocker to this PR is related to [JSDoc on the prop type definitions](https://github.com/reakit/reakit/pull/146/files#diff-57c9fc732c172670b36091d786364fd0R11). With a normal react component these will show up in [VSCode autocomplete intellisense](https://code.visualstudio.com/docs/languages/javascript#_automatic-type-acquisition) and, however the way we are constructing these components by passing them through a couple of functions seems to break that.


Lastly, (also not a blocker) i think it would be desirable to continue to support `prop-types` on the reakit components as they are beneficial to people not using typescript or flow, however it would be nice to not have to duplicate type definitions in both formats. https://github.com/gcanti/prop-types-ts looks like it could be promising.

**Comment by codyaverett** [comment]
> We are close to getting the core pieces converted to typescript, those are the as helper and the Base component. Almost everything else is built on top of these so once they are in place the rest of the type conversion should go pretty smoothly crossed_fingers .

Nice! That sounds very promising.

> I did successfully get it working in the latest commit, but it was only working with styled-components type definitions from 3.3.3 and broke again with 3.4, probably not anything too complicated to figure out.

This commit? d8b36354beccdcbbbab915e5e8f6facef97df5f2

I normally transpile my code with typescript directly with `tsc` and `webpack` and I don't worry to much about babel configuration at all. I'm interested to try and fix.

What does that ideal build set up look like to you?

I'll tinker some more.  Thanks!

**Comment by wordofchristian** [comment]
> This commit? d8b3635

It was https://github.com/reakit/reakit/pull/146/commits/d66c6ef2b40d6f56cd29f09d1026220c5f8a88cd actually

It wasn't working on CI though due to the fact that the styled-components version isn't locked to 3.3.3.

If you want to see it working you'll have to install that specific version.

> I normally transpile my code with typescript directly with tsc and webpack and I don't worry to much about babel configuration at all. I'm interested to try and fix.
> 
> What does that ideal build set up look like to you?

We decided the best approach was going to be use the typescript parser that's built into babel 7 to handle the transpilation, that way we wouldn't have to introduce webpack and it would work much better with jest (which can be tricky to make work with webpack IIRC). 

Babel 7 just hit RC so hopefully it will be coming out soon.

**Comment by codyaverett** [comment]
Ah, I see.  Thanks for the explanation.  I'll try out  d66c6ef.

I personally have not had issues with Jest and webpack together. [ts-jest](https://github.com/kulshekhar/ts-jest) has worked great.  What I have heard is that there were some limitations with Jest and Create React App.  In CRA only a limited number of jest features are available unless you monkey patch your own webpack config into the build.

Anyways, I'm quite interested to see how babel7 handles with typescript.  Hopefully it's really great! :100:

**Comment by diegohaz** [comment]
I spent the last two days on this and I think it's finally done (hopefully). 

I'm gonna merge it to avoid conflicts in the near future, but it's still open for changes and improvements. So anyone reading this feel free to chime in and give feedback. Very few of the public API has been typed, so we still have a lot of work to do. XD

I'm gonna highlight here the relevant changes.

## TypeScript and JavaScript together

It would be impractical to rewrite the entire project in TS, and I don't even think that the whole codebase should be ported to TS, so our goal was to apply TS gradually and make it live together with JS.

It's possible because of babel 7 and its new typescript plugin. The build process basically follows these steps:

1. Remove all the build folders (`dist`, `es`, `lib` and the new `ts`) and proxy folders introduced in #168.
2. Use `tsc --emitDeclarationOnly --allowJs false` to copy type declarations from TS files into the `ts` folder.
3. Use `rollup` to create transpiled files in `dist`, `es` and `lib` folders. Babel 7 acts here, removing type declarations properly.
4. Create proxy folders.

It's all done with `yarn build`.

## ESLint runs both for TS and JS files

Thanks to `typescript-eslint-parser` and `eslint-plugin-typescript`, we're linting both TS and JS pretty consistently. Some rules had to be disabled for TS though.

## Singel and benchmark removed (for now)

[Singel](https://github.com/diegohaz/singel) is still not using babel 7, so it wouldn't work with our TS files. It was removed temporarily.

Benchmark was another guy that (because of `react-benchmark`) didn't work with babel 7. Also, I wasn't fully satisfied with that benchmarking approach. I'm planning to come up with a better solution soon.

## Map transpiled files to typings

When the user imports something in the library, the tooling (typescript, vscode, intellisense etc.) should know where to find the correct types for that module.

There're two main ways to import modules from ReaKit:
```js
import { Base } from "reakit"; // index.js in es or lib folder
import Base from "reakit/Base"; // proxy that maps to the file in es or lib folder
```
These other paths are also possible, although not supported (that is, recommended in the docs):
```js
import { Base } from "reakit/lib"; // accessing index.js in lib directly
import { Base } from "reakit/es"; // accessing index.js in es directly
import Base from "reakit/lib/Base"; // accessing the file inside lib directly
import Base from "reakit/es/Base"; // accessing the file inside es directly
```

Except for the last method (accessing the file directly), all the other imports are correctly mapped to their typings.

The `package.json` file inside `reakit/Base` (proxy folder) will look like this:
```json
{
  "name": "reakit/Base",
  "private": true,
  "main": "../lib/Base",
  "module": "../es/Base",
  "types": "../ts/Base"
}
```

That `types` field is where we say where the tooling will find the types for that particular module. We also have `types` in the main `package.json` and inside `lib` and `es` folders.

## File structure

I've changed the file structure to reflect better our public API. Before, we had `components`, `enhancers` and `utils` folders. Now, all the public files/folders are in `src`, and private ones start with `_` in their names:

```
src
├── __tests__
├── _utils
├── Arrow
├── Avatar
├── Backdrop
├── Base
├── Block
├── ...
├── as.tsx
├── index.ts
└── styled.ts
```

This file structure also reflects better the structure of transpiled folders (`lib`, `es`) and therefore we get the types mapping out of the box when using the first importing method.

**Comment by wordofchristian** [comment]
Thanks for pulling this over the finish line @diegohaz 🏁

**Pull request merged**